### DISCUSSION TITLE
General questions #2

### DISCUSSION THREAD

**Discussion by float34** [microsoft/WindowsAppSDK#1942] [Q&A]
Hello! 
Couple of questions.

@andrewleader has provided a good response (thank you!) in https://github.com/microsoft/WindowsAppSDK/discussions/354 (I'd like to know about performance overhead still, though ;-) )

But there are few more:
1.  Let's say I want to create a very simple Console/Desktop .NET 5/6 app, and at the same time I want to use new Windows platform capabilities (e.g. **Windows.Media.Ocr** class). 
1.1 How do I do this? Am I required to use WinAppSDK, or I can simply install a C#/WinRT language projection Nuget package and call these APIs directly?
1.2 If it's the latter- why use WinAppSDK when one can simply call WinRT APIs directly from C#/C++? (Not considering the "decoupling" aspect of WinAppSDK)
1.3 It may be that I am really dump and incompetent - let me know if so, but I feel seriously confused as to what technology to use in what scenario. How about creating a matrix, or whatever form of visual representation will work better and show ALL relations of all current Microsoft development technologies/stacks/SDKs?

2. Does using WinAppSDK or WinRT language projection make a regular Desktop app a UWP app? I want to make a desktop app which is **not** restricted as UWP app (e.g. having full file system access, not running in a sandbox, etc.) but has access to **all** modern APIs. How can I achieve this?

3. Initially I thought that WinAppSDK will encapsulate all language projections/modern platform features under one umbrella "automatically", but this does not seem to be the case. What is the relation between **WinAppSDK** and **modern APIs/language projections** now? Should one use just the first? The second? Both?

4. What about Python/WinRT? Has the development stopped? (Maybe not the best place to ask this but anyway - @kennykerr can you please clarify this one?)

Thanks!

**Comment by ghost** [comment]
**TL.DR  1 :** Don't have any presumptions. Always start from the first principles.  

**TL.DR  2  :**  You are free to ignore everything written below as unofficial response (Honestly speaking you have no other option since you yourself stated )

> 1.3 It may be that I am really dump and incompetent - let me know if so, but I feel seriously confused as to what technology to use in what scenario.

but you are also free to examine them and the fact checks yourself. while doing that, remember TL.DR 1 . you'll thank me some months later 

---


> let's say I want to create a very simple Console/Desktop .NET 5/6 app, and at the same time I want to use new Windows platform capabilities (e.g. Windows.Media.Ocr class).
> 1. Am I required to use WinAppSDK, 

No

> or I can simply install a C#/WinRT language projection Nuget package and call these APIs directly?   

Yes

> 1.2 If it's the latter- why use WinAppSDK when one can simply call WinRT APIs directly from C#/C++? (Not considering the "decoupling" aspect of WinAppSDK)

Windows App SDK is only needed when you need Sugercoated APIs(which are already available as C based aka Win32 version) and WinUI3 and WinUI3 Xaml Islands and you belong to  
**"Everything Must Be Open Sourcism or We Riot"**( https://github.com/microsoft/WindowsAppSDK/discussions/1615#discussioncomment-1875122) subsect Crowd. 


> Does using WinAppSDK or WinRT language projection make a regular Desktop app a UWP app? 

No. 

Let me handle it from first principles.

UWP = `CoreWindow` nothing less nothing more, there is no such thing as UWP APIs or UWP app, all are win32 apps and Win32 APIs.  

=> **_if you mean UWP APIs = WinRT APIs_** , then know that WinRT is sugercoat over COM, All WinRT APIs are COM APIs underneath and All COM APIs are based on C based Win32 APIs underneath. `CoreWindow` itself is built on win32 subsystem's  Window Handle (known as `hWnd` and contrary to regular win32 Apps `hWnd` ,this `CoreWindow`'s `hWnd` is hidden from devs aka system managed)  so there you go, everything is Win32 APIs underneath and everything is win32 Apps. **UWP apps** is nothing more than a buzzword from the first principles. 

=> **_if you mean UWP apps = AppContainer_** , then see https://github.com/microsoft/WindowsAppSDK/discussions/1071#discussioncomment-1012599 
AppContainer is NT Kernel feature and therefore Any Win32 app can be run in AppContainer , see [Fun with AppContainer](https://scorpiosoftware.net/2019/01/15/fun-with-appcontainers/) 

=> **_if you mean  running your app on non-Desktop platforms_**,  
it has to meet two basic Criteria 
1. your app has to call APIs which are available on that platfrom
2. your app has to run under AppContainer. 

as long as your app meets these two criterias , it will run, WindowsAppSDK will do nothing for you in this regard. 


> I want to make a desktop app which is not restricted as UWP app (e.g. having full file system access, not running in a sandbox, etc.)

restriction comes from above mentioned AppContainer.

 > but has access to all modern APIs. How can I achieve this?

there is no such thing as Modern APIs . all of these modern apis are Sugercoated APIs(which are already available as C based).

Brand new public win32 apis such as the DIrectX Agility still are C based and will always be. Brand new public win32  apis  are WIndows Version dependent. 

---

💡  Self-Reflection Question :  **_If every app is Win32 app underneath then what's the fuss since 2012, UAP, UWP and Project Reunion and so called breaking walls between UWP and Win32 ?_** 

💡 **the Answer :** 

Basically they are moving away from 
-  system managed hWnd based Window such as `CoreWindow` and Window Framework such as UWP to regular unmanaged hWnd based Window such as  `AppWindow` and Window Framwork such as WinUI3.  

- WinRT APIs (such as `Windows.Media.Ocr`) which (since created/sugercoated in 2012) assumed a `CoreWindow` Window type aka "UWP"; are being made to work with WindowsAppSDK's `AppWindow` type through `IInitializeWithWindow`. 

Open Sourcing some parts
- such as MRT to MRT Core.

**Reply by ghost** [reply]
Regarding Projection -- 

Projection is basically COM ABI Binding. For Example - You can develop your  own programming language such as "YetAnotherLanguageThatDoesTheSameAsCpp" and you can bind that syntax to COM ABI, By doing that, you can call and communicate with All the APIs (from another languages) which are exposed through the COM ABI. this is projection is. 

You should simply reference  your language nuget (the one you want to work with)  in your project and start calling APIs you want, right way.
